File: Range Sum of BST_dfs.py
#time: O(n)
#space: O(n)

#dfs


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        self.sum1=0
        def dfs(root):
            if(root==None):
                return
            if(low<=root.val<=high):
                self.sum1+=root.val
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        return self.sum1
            
            

File: Range_Sum_of_BST_DFS.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// TC : O(n)
// SC : O(n)
// Method - DFS
class Solution {
    int sum = 0;
    public int rangeSumBST(TreeNode root, int low, int high) {
        dfs(root, low, high);
        return sum;
    }
    
    public void dfs(TreeNode root, int low, int high) {
        if(root == null) return;
        
         if(root.val >= low && root.val <= high)
            sum = sum + root.val;
         if(root.val >= low)
             dfs(root.left, low, high);
         if(root.val <= high)
             dfs(root.right, low, high);
    }
}
